src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1015,73 +1005,116 @@ TargetEntryChangesValue(TargetEntry *targetEntry, Var *column, FromExpr *joinTre
 
 
 /*
- * RouterModifyTask builds a Task to represent a modification performed by
- * the provided query against the provided shard interval. This task contains
- * shard-extended deparsed SQL to be run during execution.
+ * RouterModifyTasks builds a List of Tasks representing a modification per-
+ * formed by the provided query using the provided list of modify routes.
+ * If possible, all tasks contain shard-specific SQL to be sent to workers
+ * during execution.
  */
-static Task *
-RouterModifyTask(Oid distributedTableId, Query *originalQuery,
-				 ShardInterval *shardInterval)
+static List *
+RouterModifyTasks(Oid distributedTableId, Query *query, List *modifyRouteList)
 {
-	Task *modifyTask = NULL;
+	List *modifyTasks = NIL;
 	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(distributedTableId);
+	ListCell *modifyRouteCell = NULL;
 
-	modifyTask = CitusMakeNode(Task);
-	modifyTask->jobId = INVALID_JOB_ID;
-	modifyTask->taskId = INVALID_TASK_ID;
-	modifyTask->taskType = MODIFY_TASK;
-	modifyTask->queryString = NULL;
-	modifyTask->anchorShardId = INVALID_SHARD_ID;
-	modifyTask->dependedTaskList = NIL;
-	modifyTask->replicationModel = cacheEntry->replicationModel;
-
-	if (originalQuery->onConflict != NULL)
+	foreach(modifyRouteCell, modifyRouteList)
 	{
-		RangeTblEntry *rangeTableEntry = NULL;
+		ModifyRoute *modifyRoute = (ModifyRoute *) lfirst(modifyRouteCell);
+		Task *modifyTask = CitusMakeNode(Task);
+
+		modifyTask->jobId = INVALID_JOB_ID;
+		modifyTask->taskId = INVALID_TASK_ID;
+		modifyTask->taskType = MODIFY_TASK;
+		modifyTask->queryString = NULL;
+		modifyTask->anchorShardId = INVALID_SHARD_ID;
+		modifyTask->dependedTaskList = NIL;
+		modifyTask->replicationModel = cacheEntry->replicationModel;
+		modifyTask->valuesRTEIndex = 0;
+		modifyTask->rowValuesLists = NIL;
+
+		if (query->onConflict != NULL)
+		{
+			RangeTblEntry *rangeTableEntry = NULL;
 
-		/* set the flag */
-		modifyTask->upsertQuery = true;
+			/* set the flag */
+			modifyTask->upsertQuery = true;
 
-		/* setting an alias simplifies deparsing of UPSERTs */
-		rangeTableEntry = linitial(originalQuery->rtable);
-		if (rangeTableEntry->alias == NULL)
+			/* setting an alias simplifies deparsing of UPSERTs */
+			rangeTableEntry = linitial(query->rtable);
+			if (rangeTableEntry->alias == NULL)
+			{
+				Alias *alias = makeAlias(CITUS_TABLE_ALIAS, NIL);
+				rangeTableEntry->alias = alias;
+			}
+		}
+
+		if (modifyRoute != NULL)
 		{
-			Alias *alias = makeAlias(CITUS_TABLE_ALIAS, NIL);
-			rangeTableEntry->alias = alias;
+			int64 shardId = modifyRoute->shardId;
+
+			/* grab shared metadata lock to stop concurrent placement additions */
+			LockShardDistributionMetadata(shardId, ShareLock);
+
+			modifyTask->anchorShardId = shardId;
+			modifyTask->valuesRTEIndex = modifyRoute->partitionvarno;
+			modifyTask->rowValuesLists = modifyRoute->rowValuesLists;
```

Comment:
Can we avoid storing the `partitionvarno` and `rowValuesList` in `Task` by passing it to `UpdateTaskQueryString`? After we do the deparsing in `UpdateTaskQueryString` we don't need the routes anymore. When we defer pruning to the executor we rebuild the list of routes anyway.

I guess the case in which storing them might be beneficial is when the partition column values are known, but some other columns contain function calls. In that case we can build the list of tasks, but we cannot deparse them yet. On the other hand, we're deparsing them on the line below, so we'll actually do deparsing twice, which is really expensive. It seems simpler and equally performant to just build the tasks in the executor after function evaluation in case there are function calls.